osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs

```diff
@@ -154,13 +155,17 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             }
 
             var result = ResultFor(timeOffset);
+            var clickAction = CheckHittable?.Invoke(this, Time.Current);
 
-            if (result == HitResult.None || CheckHittable?.Invoke(this, Time.Current) == false)
+            if (clickAction == ClickAction.Shake || (result == HitResult.None && clickAction != ClickAction.Ignore))
             {
                 Shake();
                 return;
             }
 
+            if (result == HitResult.None)
+                return;
+
```

Comment:
this is the part I'm most skeptical of. If the hit result is `HitResult.None`, we still need to check `clickAction`: a hit could be out of a note's hitwindow (and otherwise be shaken), but the hit policy says that it should be ignored, not shaken. 

Also, the hit policy could theoretically say that the hitobj should be shaken even if the hit result isn't None, although I don't think it does right now because stable notelock uses the same hitwindow as judgment results (400ms for misses). That is the cause for the first condition.

Is there a cleaner way to do all this?

Reply:
I mean, I don't _think_ there is any way to clean this up, because what these conditionals essentially are, is legacy crap leaking out of `IHitPolicy`.

_Maybe_ if you shoved `result` and the `None`-checking stuff into the hit policy you could move this ugliness out of the way a _tiny teeny little bit_. But I'm not sure it's worth the payoff.

Reply:
right, it's the leakage from `IHitPolicy` that I was thinking most about. Maybe there's some way to isolate the effects of `ClickAction.Ignore` to `ObjectOrderedHitPolicy` instead of `IHitPolicy`, which complicates all other hit policies as well. But I don't see an easy way out.

Reply:
What I had in mind is basically (warning: very untested)

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs b/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs
index c818a361df..af02087d1a 100644
--- a/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs
+++ b/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs
@@ -131,7 +131,7 @@ public TestDrawableHitCircle(HitCircle h, bool auto, double hitOffset)
 
             protected override void CheckForResult(bool userTriggered, double timeOffset)
             {
-                if (auto && !userTriggered && timeOffset > hitOffset && CheckHittable?.Invoke(this, Time.Current) == ClickAction.Hit)
+                if (auto && !userTriggered && timeOffset > hitOffset && CheckHittable?.Invoke(this, Time.Current, HitResult.Great) == ClickAction.Hit)
                 {
                     // force success
                     ApplyResult(r => r.Type = HitResult.Great);
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs
index 09d818def8..eb56b21684 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs
@@ -155,13 +155,10 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             }
 
             var result = ResultFor(timeOffset);
-            var clickAction = CheckHittable?.Invoke(this, Time.Current);
+            var clickAction = CheckHittable?.Invoke(this, Time.Current, result);
 
-            if (clickAction == ClickAction.Shake || (result == HitResult.None && clickAction != ClickAction.Ignore))
-            {
+            if (clickAction == ClickAction.Shake)
                 Shake();
-                return;
-            }
 
             if (result == HitResult.None)
                 return;
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
index a8ce2118c8..920dfcab03 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
@@ -13,6 +13,7 @@
 using osu.Game.Rulesets.Osu.Judgements;
 using osu.Game.Rulesets.Osu.Scoring;
 using osu.Game.Rulesets.Osu.UI;
+using osu.Game.Rulesets.Scoring;
 using osuTK;
 using osuTK.Graphics;
 
@@ -37,7 +38,7 @@ public abstract partial class DrawableOsuHitObject : DrawableHitObject<OsuHitObj
         /// and <see cref="ClickAction.Shake"/>, and this hit object will be shaken for return values of
         /// <see cref="ClickAction.Shake"/>.
         /// </summary>
-        public Func<DrawableHitObject, double, ClickAction> CheckHittable;
+        public Func<DrawableHitObject, double, HitResult, ClickAction> CheckHittable;
 
         protected DrawableOsuHitObject(OsuHitObject hitObject)
             : base(hitObject)
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderHead.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderHead.cs
index a4cf69ee31..41f6a40c0a 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderHead.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderHead.cs
@@ -61,7 +61,7 @@ protected override void OnApply()
 
             pathVersion.BindTo(DrawableSlider.PathVersion);
 
-            CheckHittable = (d, t) => DrawableSlider.CheckHittable?.Invoke(d, t) ?? ClickAction.Hit;
+            CheckHittable = (d, t, r) => DrawableSlider.CheckHittable?.Invoke(d, t, r) ?? ClickAction.Hit;
         }
 
         protected override void Update()
diff --git a/osu.Game.Rulesets.Osu/UI/AnyOrderHitPolicy.cs b/osu.Game.Rulesets.Osu/UI/AnyOrderHitPolicy.cs
index 7503c43e0b..69bd360b5e 100644
--- a/osu.Game.Rulesets.Osu/UI/AnyOrderHitPolicy.cs
+++ b/osu.Game.Rulesets.Osu/UI/AnyOrderHitPolicy.cs
@@ -4,6 +4,7 @@
 #nullable disable
 
 using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Scoring;
 using osu.Game.Rulesets.UI;
 
 namespace osu.Game.Rulesets.Osu.UI
@@ -15,7 +16,7 @@ public class AnyOrderHitPolicy : IHitPolicy
     {
         public IHitObjectContainer HitObjectContainer { get; set; }
 
-        public ClickAction CheckHittable(DrawableHitObject hitObject, double time) => ClickAction.Hit;
+        public ClickAction CheckHittable(DrawableHitObject hitObject, double time, HitResult result) => ClickAction.Hit;
 
         public void HandleHit(DrawableHitObject hitObject)
         {
diff --git a/osu.Game.Rulesets.Osu/UI/IHitPolicy.cs b/osu.Game.Rulesets.Osu/UI/IHitPolicy.cs
index 9820b8c188..44d3b37408 100644
--- a/osu.Game.Rulesets.Osu/UI/IHitPolicy.cs
+++ b/osu.Game.Rulesets.Osu/UI/IHitPolicy.cs
@@ -3,6 +3,7 @@
 
 using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Scoring;
 using osu.Game.Rulesets.UI;
 
 namespace osu.Game.Rulesets.Osu.UI
@@ -19,8 +20,9 @@ public interface IHitPolicy
         /// </summary>
         /// <param name="hitObject">The <see cref="DrawableHitObject"/> to check.</param>
         /// <param name="time">The time to check.</param>
+        /// <param name="result">The result that the object would be judged with if hit.</param>
         /// <returns>Whether <paramref name="hitObject"/> can be hit at the given <paramref name="time"/>.</returns>
-        ClickAction CheckHittable(DrawableHitObject hitObject, double time);
+        ClickAction CheckHittable(DrawableHitObject hitObject, double time, HitResult result);
 
         /// <summary>
         /// Handles a <see cref="HitObject"/> being hit.
diff --git a/osu.Game.Rulesets.Osu/UI/LegacyHitPolicy.cs b/osu.Game.Rulesets.Osu/UI/LegacyHitPolicy.cs
index c35d4a1b56..4c09ca630a 100644
--- a/osu.Game.Rulesets.Osu/UI/LegacyHitPolicy.cs
+++ b/osu.Game.Rulesets.Osu/UI/LegacyHitPolicy.cs
@@ -8,6 +8,7 @@
 using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Drawables;
 using osu.Game.Rulesets.Osu.Objects.Drawables;
+using osu.Game.Rulesets.Scoring;
 using osu.Game.Rulesets.UI;
 
 namespace osu.Game.Rulesets.Osu.UI
@@ -26,7 +27,7 @@ public void HandleHit(DrawableHitObject hitObject)
         {
         }
 
-        public ClickAction CheckHittable(DrawableHitObject hitObject, double time)
+        public ClickAction CheckHittable(DrawableHitObject hitObject, double time, HitResult result)
         {
             var aliveObjects = HitObjectContainer.AliveObjects.ToList();
             int index = aliveObjects.IndexOf(hitObject);
@@ -38,6 +39,9 @@ public ClickAction CheckHittable(DrawableHitObject hitObject, double time)
                     return ClickAction.Ignore;
             }
 
+            if (result == HitResult.None)
+                return ClickAction.Shake;
+
             foreach (DrawableHitObject testObject in aliveObjects)
             {
                 if (testObject.AllJudged)
diff --git a/osu.Game.Rulesets.Osu/UI/StartTimeOrderedHitPolicy.cs b/osu.Game.Rulesets.Osu/UI/StartTimeOrderedHitPolicy.cs
index f33ca58aef..f647918bd1 100644
--- a/osu.Game.Rulesets.Osu/UI/StartTimeOrderedHitPolicy.cs
+++ b/osu.Game.Rulesets.Osu/UI/StartTimeOrderedHitPolicy.cs
@@ -8,6 +8,7 @@
 using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Drawables;
 using osu.Game.Rulesets.Osu.Objects.Drawables;
+using osu.Game.Rulesets.Scoring;
 using osu.Game.Rulesets.UI;
 
 namespace osu.Game.Rulesets.Osu.UI
@@ -24,7 +25,7 @@ public class StartTimeOrderedHitPolicy : IHitPolicy
     {
         public IHitObjectContainer HitObjectContainer { get; set; }
 
-        public ClickAction CheckHittable(DrawableHitObject hitObject, double time)
+        public ClickAction CheckHittable(DrawableHitObject hitObject, double time, HitResult hitResult)
         {
             DrawableHitObject blockingObject = null;
 
@@ -51,7 +52,7 @@ public void HandleHit(DrawableHitObject hitObject)
             if (!hitObjectCanBlockFutureHits(hitObject))
                 return;
 
-            if (CheckHittable(hitObject, hitObject.HitObject.StartTime + hitObject.Result.TimeOffset) != ClickAction.Hit)
+            if (CheckHittable(hitObject, hitObject.HitObject.StartTime + hitObject.Result.TimeOffset, hitObject.Result.Type) != ClickAction.Hit)
                 throw new InvalidOperationException($"A {hitObject} was hit before it became hittable!");
 
             // Miss all hitobjects prior to the hit one.

```

It gets rid of the ugly conditional but makes `IHitPolicy` ugly instead. Dunno if worth the tradeoff.